### PR TITLE
Define task source, construct with realms, improve parallel usage

### PR DESCRIPTION
**Pull request by inexorabletash**
- Use a more common phrasing for "in parallel" steps.
- Make "in parallel" operations interact with script via queued tasks.
- Define an "ml task source" used for queued tasks.
- Ensure the correct Realm is used for asynch construction.

Relevant requirements/examples:
- https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel
- https://webidl.spec.whatwg.org/#js-promise-examples

***
<a href="https://pr-preview.s3.amazonaws.com/inexorabletash/webnn/pull/593.html" title="Last updated on Mar 5, 2024, 8:24 PM UTC (8510ffc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/593/ca0b903...inexorabletash:8510ffc.html" title="Last updated on Mar 5, 2024, 8:24 PM UTC (8510ffc)">Diff</a>

### CODE REVIEW

**Comment by inexorabletash** [comment]
@reillyeon @a-sully - can you take an initial look?

One thing I didn't touch that should probably be improved: "execute graph" still throws (DOMException, TypeError) which shouldn't happen in parallel, even though compute() takes that "throw" and turns it into a reject.

**Comment by a-sully** [comment]
> One thing I didn't touch that should probably be improved: "execute graph" still throws (DOMException, TypeError) which shouldn't happen in parallel, even though compute() takes that "throw" and turns it into a reject.

FWIW the [File System spec](https://fs.spec.whatwg.org/) solves this by [making the respective algorithm optionally return a `DOMException` error name](https://fs.spec.whatwg.org/#file-system-access-result-error-name) and [expecting the caller to queue a task to reject, as appropriate](https://fs.spec.whatwg.org/#:~:text=Dependent%20specifications%20may,the%20empty%20string.)

File: index.bs [hunk 1]
```diff
@@ -835,11 +846,14 @@ The <dfn>power preference</dfn> indicates preference as related to power consump
     The <dfn method for=ML>createContext(|options|)</dfn> steps are:
 </summary>
 <div class=algorithm-steps>
-    1. If [=this=]'s [=relevant global object=]'s [=associated Document=] is not [=allowed to use=] the [=webnn-feature|webnn=] feature, return [=a new promise=] [=rejected=] with a "{{SecurityError}}" {{DOMException}}.
+    1. Let |global| be [=this=]'s [=relevant global object=].
+    1. If |global|'s [=associated Document=] is not [=allowed to use=] the [=webnn-feature|webnn=] feature, return [=a new promise=] [=rejected=] with a "{{SecurityError}}" {{DOMException}}.
+    1. Let |realm| be [=this=]'s [=relevant realm=].
     1. Let |promise| be [=a new promise=].
```
Comment (a-sully) [inline index.bs:852]: optional cleanup here and below: declare `|promise|` before checking the webnn-feature, as to not repeat creating `[=a new promise=]` (could also defer this to a follow-up, if you'd like)

**Review by a-sully — APPROVED** [review]
LGTM! I'm happy with deferring the "execute graph" problem to a follow-up

**Review by reillyeon — APPROVED** [review]
I have only one question which is whether it should be "ml task source" or "ML task source".

File: index.bs [hunk 2]
```diff
@@ -736,6 +738,15 @@ In a situation when a GPU context executes a graph with a constant or an input i
 
 When an {{MLContext}} is created with {{MLContextOptions}}, the user agent selects and creates the underlying execution device by taking into account the application's [=power preference=] and [=device type=] specified in the {{MLPowerPreference}} and {{MLDeviceType}} options.
 
+## Task Source ## {#programming-model-task-source}
+
+The <dfn>ml task source</dfn> is a [=task source=] to be used for all [=tasks=] related to asynchronous compilation and execution of {{MLGraph}}s.
```
Comment (huningxin) [inline index.bs:743]: "ml task source" seems also to be used for context creation, should it also be mentioned here?

**Review by huningxin — APPROVED** [review]
LGTM with a question, thanks much!

**Comment by inexorabletash** [comment]
> I have only one question which is whether it should be "ml task source" or "ML task source".

Capitalized "ml task" to "ML task" everywhere.

File: index.bs (hunk 2 above)
Comment (inexorabletash) [reply index.bs]: Added!

File: index.bs (hunk 1 above)
Comment (inexorabletash) [reply index.bs:852]: I'll defer, and FWIW I prefer the current structure where `|promise|` is only used for the "in parallel" steps, and the almost-synchronous failures are handled in a single line of the spec.

It also doesn't save text:

* If ... then return a new promise rejected with a ....
* If ... then reject promise with a ... and return promise.

Comment (a-sully) [reply index.bs:852]: SGTM 👍

**Comment by inexorabletash** [comment]
> > One thing I didn't touch that should probably be improved: "execute graph" still throws (DOMException, TypeError) which shouldn't happen in parallel, even though compute() takes that "throw" and turns it into a reject.
> 
> FWIW the [File System spec](https://fs.spec.whatwg.org/) solves this by [making the respective algorithm optionally return a `DOMException` error name](https://fs.spec.whatwg.org/#file-system-access-result-error-name) and [expecting the caller to queue a task to reject, as appropriate](https://fs.spec.whatwg.org/#:~:text=Dependent%20specifications%20may,the%20empty%20string.)

I decided to tackle this and made a slight variation - the steps now return an error, and if an error is returned the invoker rejects with an equivalent error in the right realm. I think that's probably precise enough without being too wordy. I didn't just use an error name because TypeError and "OperationError" DOMException are both used.

**Comment by a-sully** [comment]
> > > One thing I didn't touch that should probably be improved: "execute graph" still throws (DOMException, TypeError) which shouldn't happen in parallel, even though compute() takes that "throw" and turns it into a reject.
> > 
> > 
> > FWIW the [File System spec](https://fs.spec.whatwg.org/) solves this by [making the respective algorithm optionally return a `DOMException` error name](https://fs.spec.whatwg.org/#file-system-access-result-error-name) and [expecting the caller to queue a task to reject, as appropriate](https://fs.spec.whatwg.org/#:~:text=Dependent%20specifications%20may,the%20empty%20string.)
> 
> I decided to tackle this and made a slight variation - the steps now return an error, and if an error is returned the invoker rejects with an equivalent error in the right realm. I think that's probably precise enough without being too wordy. I didn't just use an error name because TypeError and "OperationError" DOMException are both used.

Seems reasonable to me 👍

File: index.bs [hunk 3]
```diff
@@ -980,20 +997,21 @@ In accordance with the [=ArrayBufferView/write|Web IDL warning=], to prevent the
...
+    1. Let |global| be [=this=]'s [=relevant global object=].
+    1. Let |realm| be [=this=]'s [=relevant realm=].
+    1. Run the following steps [=in parallel=]:
+        1. If |graph|.{{MLGraph/[[context]]}} is not [=this=], then [=queue an ML task=] with |global| to [=reject=] |promise| with a {{TypeError}} and abort these steps.
+        1. If |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[contextType]]}} is not "[=context type/default=]", then [=queue an ML task=] with |global| to [=reject=] |promise| with an "{{OperationError}}" {{DOMException}} and abort these steps.
+        1. If [=validating graph resources=] given |inputs| and |graph|.{{MLGraph/[[inputDescriptors]]}} returns false, then [=queue an ML task=] with |global| to [=reject=] |promise| with a {{TypeError}} and abort these steps.
+        1. If [=validating graph resources=] given |outputs| and |graph|.{{MLGraph/[[outputDescriptors]]}} returns false, then [=queue an ML task=] with |global| to [=reject=] |promise| with a {{TypeError}} and abort these steps.
+        1. Let |transferredInputs| be the result of [=MLNamedArrayBufferViews/transfer|transferring=] {{MLNamedArrayBufferViews}} |inputs| with |realm|.
```
Comment (reillyeon) [inline index.bs:1007]: As a follow-up, I don't think you can transfer a buffer during "in parallel" because that would change the state of the buffer in parallel with script execution.

Most of these steps in fact should happen before going "in parallel".

**Review by reillyeon — APPROVED** [review]

File: index.bs (hunk 3 above)
Comment (inexorabletash) [reply index.bs:1007]: Doing it here would simplify the steps, so... at the risk of feature creep...

Comment (inexorabletash) [reply index.bs]: Updated to move the steps out of the "in parallel"

**Review by reillyeon — APPROVED** [review]

**Review by huningxin — APPROVED** [review]
LGTM!

**Review by fdwr — APPROVED** [review]
